### ISSUE TITLE
Implement ITF

### ISSUE THREAD

**Issue by pbhofstede** [Spelt/ZXing.Delphi#2]
Whow, I love your work already! No binaries needed if you want to deploy on android or ios. QR-code scanning is working like a charm.
The ZXing barcode app can scan my ITF-barcode. Can we implement this together in ZXing.Delphi?
If you can decompile it to Delphi.net I will try to convert it to Delphi FMX/VCL

**Comment by Spelt** [comment]
Thanks!

Sounds like fun.  I'll send you the decompiled code in a few hours.

**Comment by Spelt** [comment]
Here it is! It should be one unit. All base classes are I think already implemented. 


In Delphi classes start with a T so the the class declaration should be:

public TITFReader = class sealed (TOneDReader)

etc etc.

Other things to consider:

- Start with adding a test in the unit test. The type ITF already exists. 
- I make use of generic arrays.  
- if there is a shr it must be replaced with TMathUtils.Asr because a .NET shift right is not the same as a Delphi shift right.
- You must supply a unit test methode with at least 3 barcode examples. The more the better. QR Code was complicated so there had to be lots of tests. Just add in Test.pas. Add also a few in the ‘Auto’ test. The new barcode should probably added to the ‘Auto’ clause.


Also take a look at the code96reader.pas or the code128reader.pas for a ‘How I did it’ help.



for reference the original source:  https://github.com/Redth/ZXing.Net.Mobile/blob/master/Source/ZXing.Net/oned/ITFReader.cs <https://github.com/Redth/ZXing.Net.Mobile/blob/master/Source/ZXing.Net/oned/ITFReader.cs> 


Some words from the original author:
Implements decoding of the ITF format, or Interleaved Two of Five.
This Reader will scan ITF barcodes of certain lengths only. At the moment it reads length 6, 8, 10, 12, 14, 16, 18, 20, 24, 44 and 48 as these have appeared "in the wild". Not all lengths are scanned, especially shorter ones, to avoid false positives. This in turn is due to a lack of required checksum function.
The checksum is optional and is not applied by this Reader. The consumer of the decoded value will have to apply a checksum if required.
http://en.wikipedia.org/wiki/Interleaved_2_of_5 <http://en.wikipedia.org/wiki/Interleaved_2_of_5> is a great reference for Interleaved 2 of 5 information.

Hope it helps !

Edward

**Comment by Spelt** [comment]
looks like the upload of the file failed. I tried it again and it seems to work.

**Comment by Spelt** [comment]
Did you have any progress?

**Comment by pbhofstede** [comment]
Hi,

Yeah, the conversion went well and I managed to scan the barcodes.
After testing the project was set on halt so wasn't digging further.

If you wish I can send you the code. I havent suppleid it with unit tests.

grtz

Date: Fri, 29 Jan 2016 03:12:01 -0800

**Comment by Spelt** [comment]
Sure!  Send it over. I'll add some tests.

**Comment by pbhofstede** [comment]
You should have got the pas-file in the prev. mail.


Date: Fri, 29 Jan 2016 04:21:16 -0800

**Comment by Spelt** [comment]
Unfortunately, I did not receive the pas file.

2016-01-29 13:22 GMT+01:00 pbhofstede <notifications@github.com>:

**Comment by pbhofstede** [comment]
Did you received it now?

Date: Fri, 29 Jan 2016 04:57:42 -0800

**Comment by Spelt** [comment]
nothing yet, maybe zip it. It probably get filtered by github.

2016-01-29 13:59 GMT+01:00 pbhofstede <notifications@github.com>:

**Comment by pbhofstede** [comment]
here you are

Date: Fri, 29 Jan 2016 05:04:08 -0800

**Comment by Spelt** [comment]
ah I see that it cant be done. You need to attach it in the issue via the
browser :-(
https://github.com/Spelt/ZXing.Delphi/issues/2

Files supported:
https://help.github.com/articles/file-attachments-on-issues-and-pull-requests/


2016-01-29 14:05 GMT+01:00 pbhofstede <notifications@github.com>:

**Comment by pbhofstede** [comment]
Whaha, I know what happened. I just replied your mail while I wasn't aware your mail was auto generated by github.
I've created a pull request 

Date: Fri, 29 Jan 2016 05:12:43 -0800

**Comment by Spelt** [comment]
not seeing any pull requests :-(

**Comment by pbhofstede** [comment]
Even not on:  https://github.com/Spelt/ZXing.Delphi/pulls
If you can leave a mail address I can send it by mail.

Date: Fri, 29 Jan 2016 05:31:08 -0800

**Comment by Spelt** [comment]
Hi

I added the ITF + unit tests to the library. I did a minor fix in your code
and added some array bounds protection code. This was needed otherwise
there was a problem in 'auto mode' with QR codes.

I even discovered a small bug in a base class. It probably had to do with
borders.  Maybe you deviated from the code in order to compensate for this
bug.

Anyway it works! Another barcode added.

Thanks!

2016-01-29 14:46 GMT+01:00 pbhofstede <notifications@github.com>:
